### PR TITLE
[Sema][AST][WIP] Support existentials with concrete assoc. types

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
Resolving [SR-4758](https://bugs.swift.org/browse/SR-4758).

This still needs corrections and I might have touched ABI sensitive stuff I shouldn't have, I'm not sure.

Anyway, I would like to hear your overall opinion, @slavapestov. Please be strict 🙂

Happy new year!

### TODO
- [x] Handle the edge case `Assoc.Assoc == Concrete`. 
- [x] Handle concrete associated type references in expression position, i.e.
  * `print(Proto.ConcreteAssoc.self)`, 
  * `print((A & B).ConcreteAssoc.self)`
  * `print(Proto.TypeAliasReferencingConcreteAssoc.self)`
  * ```swift
    func foo(meta1: Proto.Protocol, meta2: (A & B).Protocol) {
      print(meta1.ConcreteAssocOrTypeAliasWithTypeParameters.self)
      print(meta2.ConcreteAssocOrTypeAliasWithTypeParameters.self)
    }
    ```
- [x] Make sure we cannot access type aliases that contain `Self`.
- [x] Make sure compositions with conflicting concrete associated types get diagnosed.
- [ ] Decide on how to avoid creating generic environments for composition types each time a nested type has to be resolved.
- [ ] Handle the edge case `Assoc == Self`.

### CODE REVIEW

File: include/swift/AST/Decl.h [hunk 1]
```diff
@@ -3859,6 +3859,32 @@ struct SelfReferenceKind {
       other(other) { }
 };
 
+/// Used to optimize algorithms that compute support kinds
+/// and provide better diagnostics on the reason an existential
+/// type is not supported. For example, the knowledge that an
+/// inherited protocol is unsupported due to a \c Self reference
+/// saves us from searching for associated types and same-type
+/// constraints.
+enum class ExistentialSupportKind {
```
Comment (slavapestov) [inline include/swift/AST/Decl.h]: Can you split off a change that introduces this enum into a separate commit? It would be good to have separate diagnostics for this too, so it tells you which one of "Self or associated type requirements" the protocol actually has

File: lib/AST/Decl.cpp [hunk 2]
```diff
@@ -4087,46 +4087,152 @@ bool ProtocolDecl::isAvailableInExistential(const ValueDecl *decl) const {
...
+                     AssociatedTypeFreedomMap &map) const {
+  for (auto *proto: getInheritedProtocols())
+    proto->getAssociatedTypeFreedomMapRec(map);
+
+  for (auto *assoc: getAssociatedTypeMembers())
+    map.insert({assoc, false});
+
+  for (auto &req : getRequirementSignature()) {
+    if (req.getKind() == RequirementKind::SameType &&
+        req.getSecondType()->getKind() != TypeKind::DependentMember) {
```
Comment (slavapestov) [inline lib/AST/Decl.cpp]: This is not sufficient. The second type might be a generic type parameter type (eg, 'Self').

File: lib/AST/Decl.cpp [hunk 3]
```diff
@@ -4087,46 +4087,152 @@ bool ProtocolDecl::isAvailableInExistential(const ValueDecl *decl) const {
...
+
+      if (auto assoc = req.getFirstType()->getAs<DependentMemberType>()) {
+        auto *assocDecl = assoc->getAssocType();
```
Comment (slavapestov) [inline lib/AST/Decl.cpp]: ... the *base* type might not be Self here. For example,

```
protocol Q {
  associatedtype T
  associatedtype U where T : Q, T.T == Int, T.U == Int
}
```

I think your algorithm will flag this protocol as having both associated types fully constrained when they're not.

Instead you should look at the Self archetype of the protocol and iterate over its nested types:

```
auto selfType = protoDecl->getSelfInterfaceType();
auto archetypeType = protoDecl->mapTypeIntoContext(selfType)->castTo<ArchetypeType>()
for (auto nested : archetypeType->getNestedTypes()) {
  ...
}
```

File: lib/AST/Decl.cpp [hunk 4]
```diff
@@ -4087,46 +4087,152 @@ bool ProtocolDecl::isAvailableInExistential(const ValueDecl *decl) const {
...
+                        valueMember, /*allowCovariantParameters=*/true,
+                        /*skipAssocTypes=*/true);
+
+      // A protocol with a Self reference in contravariant position
+      // cannot be used as an existential, alongside any inheriting
+      // protocol; compute the rest of the map if needed and bail out.
+      if (selfKind.parameter || selfKind.other) {
+        if (needsCompleteAssocMap)
+          getAssociatedTypeFreedomMapRec(map);
+        setSupportKind(ExistentialSupportKind::UnsupportedSelf);
```
Comment (slavapestov) [inline lib/AST/Decl.cpp]: Set the support kind once in the caller, instead of setting it at every 'return' location here

File: lib/AST/Decl.cpp [hunk 5]
```diff
@@ -4087,46 +4087,152 @@ bool ProtocolDecl::isAvailableInExistential(const ValueDecl *decl) const {
...
-        return false;
+      auto selfKind = findProtocolSelfReferences(
+                        valueMember, /*allowCovariantParameters=*/true,
+                        /*skipAssocTypes=*/true);
+
+      // A protocol with a Self reference in contravariant position
+      // cannot be used as an existential, alongside any inheriting
+      // protocol; compute the rest of the map if needed and bail out.
+      if (selfKind.parameter || selfKind.other) {
+        if (needsCompleteAssocMap)
```
Comment (slavapestov) [inline lib/AST/Decl.cpp]: It would be better to compute the map as needed and cache it in the protocol, instead of passing this boolean down here.

File: lib/AST/GenericSignatureBuilder.cpp [hunk 6]
```diff
@@ -2990,9 +2990,36 @@ PotentialArchetype *PotentialArchetype::updateNestedTypeForConformance(
...
+      builder.addGenericParameter(firstProto->getProtocolSelfType());
+      interfaceType = firstProto->getProtocolSelfType();
+
+      for (auto *proto: layout.getProtocols()) {
+        for (auto &req : proto->getDecl()->getRequirementSignature())
+          builder.addRequirement(req, FloatingRequirementSource::forAbstract(),
+                                 nullptr);
+      }
+      auto newSig = std::move(builder).computeGenericSignature(SourceLoc());
+      genericEnv = newSig->createGenericEnvironment();
```
Comment (slavapestov) [inline lib/AST/GenericSignatureBuilder.cpp]: This looks inefficient since you're creating a lot of generic environments. They live forever and do not get deallocated. Since the constrained type is fully concrete, it must have been made fully concrete in one of the inherited protocols, so it should be possible to find that protocol without building a whole new GSB here.

File: lib/Sema/ConstraintSystem.cpp [hunk 7]
```diff
@@ -1359,24 +1359,24 @@ ConstraintSystem::getTypeOfMemberReference(
...
-      if (t->is<TypeVariableType>())
-        if (t->isEqual(selfTy))
-          return baseObjTy;
-      if (auto *metatypeTy = t->getAs<MetatypeType>())
-        if (metatypeTy->getInstanceType()->isEqual(selfTy))
-          return ExistentialMetatypeType::get(baseObjTy);
-      return t;
-    });
-  }
+//  if (!isDynamicResult &&
```
Comment (slavapestov) [inline lib/Sema/ConstraintSystem.cpp]: Was this causing problems?

File: lib/Sema/TypeCheckType.cpp [hunk 8]
```diff
@@ -3010,7 +3010,10 @@ Type TypeResolver::resolveCompositionType(CompositionTypeRepr *repr,
 
   for (auto tyR : repr->getTypes()) {
     Type ty = resolveType(tyR, options.withoutContext());
-    if (!ty || ty->hasError()) return ty;
+    if (!ty || ty->hasError()) {
+      repr->setResolvedType(ty);
```
Comment (slavapestov) [inline lib/Sema/TypeCheckType.cpp]: If this has some effect, can you split it off into a separate commit with a test case?

File: test/type/protocol_types.swift [hunk 9]
```diff
@@ -138,3 +138,54 @@ struct BadSubscript {
     set {}
   }
 }
+
+
+protocol TrivialProto {}
```
Comment (slavapestov) [inline test/type/protocol_types.swift]: Suggestions for more tests:

- Make sure that nested types of existentials work in both type position (eg, the type of a parameter) and expression position (eg, print(Foo.Bar.self)).

- Add a SILGen test file to make sure SILGen generates things correctly (take the subset of your 'valid' test cases from test/type/protocol_types.swift and put them in a SILGen test).

There's also a fully edge case involving Self:

```
protocol P {
  associatedtype A

  func foo() -> A
}

protocol Q : P where A == Self {}
```

it should be possible to call `foo()` on an existential of type Q.

**Review by slavapestov — COMMENTED** [review]
This is a great start! My main suggestion is to try to split this sort of change into smaller commits. It creates more work in the short term, but it's easier to review, test and debug this sort of change if its split into smaller pieces.

**Comment by AnthonyLatsis** [comment]
> FWIW I usually run build-script with -r since it builds a release swiftc which builds the stdlib a lot faster. Then I don't run build-script again, and instead run ninja with various targets inside the build directory. For example, ninja swift to just rebuild swiftc, or ninja swift-stdlib to rebuild the stdlib (takes a few minutes with release compiler).

Yes, that's exactly what I do. I actually tried guessing a suitable target with `ninja stdlib`. Thanks for the tip on the right spelling! My mistake was looking for it in the `build/bin` directory; now I realize the stdlib doesn't need an executable.  
**FYI** Before reading your reply, I managed to solve this by running `build-script` with `--build-swift-static-stdlib`.

File: lib/AST/Decl.cpp (hunk 5 above)
Comment (AnthonyLatsis) [reply lib/AST/Decl.cpp]: Caching is a good idea, since we only ever need fully computed maps for composition types, but isn't caching orthogonal to that boolean? It's purpose is to simply optimize the computation of the map. I do agree it looks a bit odd though.

Fortunately I found another rather trivial reason to always compute maps and support kinds separately (and thus get rid of that boolean) - if all protocols of a composition type are individually supported, we don't need maps to begin with.

**Comment by slavapestov** [comment]
@AnthonyLatsis FWIW I'm still really keen in seeing this change land eventually. Let me know if you need help

**Comment by AnthonyLatsis** [comment]
Hi Slava, I got backed up with other things I had to do these past weeks, sorry for postponing the review. I'll address all your comments in a couple hours and I am planning to get started again not long after. Naturally, I'll let you know if I have any questions. Thanks!

File: lib/AST/Decl.cpp (hunk 4 above)
Comment (AnthonyLatsis) [reply lib/AST/Decl.cpp]: The function (`existentialTypeSupportedSlow`) is recursive, so if I set the support kind only in the first caller, protocols up the hierarchy will not have their support kinds set if the function happens to recurse. This wasn't a problem before because the function called `existentialTypeSupported` instead of directly recursing, but now `existentialTypeSupportedSlow` has a new map parameter. I don't think retaining an optional map parameter on `existentialTypeSupported` too just for the purpose of setting the support kind once is reasonable.

File: lib/Sema/ConstraintSystem.cpp (hunk 7 above)
Comment (AnthonyLatsis) [reply lib/Sema/ConstraintSystem.cpp]: This piece of code was replacing `Self` with the protocol type, preventing the constraint system from treating them as archetypes and trying to resolve a nested type that can be resolved. So nothing like `resolveNestedType` or `simplifyType` was called and the compiler consistently tripped on an unhandled coercion every time there was such a nested type. I commented it out as a hack because I didn't know how else to convince the constraint system to pay attention. Here's where I need some help.

File: lib/Sema/TypeCheckType.cpp (hunk 8 above)
Comment (AnthonyLatsis) [reply lib/Sema/TypeCheckType.cpp]: Type representations do not store the types they represent. I suspect the chief reason is redundancy. In my case having such an ability for `CompositionTypeRepr` was important because otherwise I would have to duplicate the logic in `resolveCompositionType` to get ahold of a type that I can use to inspect existential support in `UnsupportedProtocolVisitor::visitCompositionTypeRepr`.

File: include/swift/AST/Decl.h (hunk 1 above)
Comment (AnthonyLatsis) [reply include/swift/AST/Decl.h]: Sure, I was going to separate improvements to diagnostics later into a separate PR, since there will probably be a lot of test corrections plus new tests, but we can append them here as well.
To clarify, you want me to separate just the enum declaration, or all the code that now depends on it?

File: lib/AST/Decl.cpp (hunk 3 above)
Comment (AnthonyLatsis) [reply lib/AST/Decl.cpp]: > ```swift
> protocol Q {
>  associatedtype T
>  associatedtype U where T : Q, T.T == Int, T.U == Int
> }
> ```
I didn't know associated types accept constraints that aren't related to them. I can't seem to make it compile with this conformance:
```swift
extension Int: Q {
  typealias T = Int
  typealias U = Int
}
```

I believe what you meant here and in the previous [comment](https://github.com/apple/swift/pull/21576#discussion_r244833860) is dealt with right after by checking whether `req.getFirstType()` corresponds to an associated type declaration *present* in the map:
```cpp
 if (map.count(assocDecl))
   map[assocDecl] = true;
```
Somehow `Assoc.Assoc` is always a different `AssociatedTypeDecl` than `Assoc`. For example, this works. 
```swift
protocol P { associatedtype Q: P }
protocol PP: P where Q.Q == Int {}

func foo2(arg: PP) {} // protocol 'PP' can only be used ...
```
Or... is `getAssocType()` just giving me a different reference? But all other cases work fine. Could it be giving me a different reference only when it's a nested associated type?

File: test/type/protocol_types.swift (hunk 9 above)
Comment (AnthonyLatsis) [reply test/type/protocol_types.swift]: > There's also a fully edge case involving Self

Yeah, `Self` is more tricky. We need a function that recursively searches for a certain associated type in covariant position, in every member up till the protocol that introduces the associated type. We could probably adapt `findProtocolSelfReferences`. But the worst part is we would have to sometimes repeat this for compositions, i.e. for `P1` in `P1(A) & P2(A == Self)`. And to do that we have to propagate the knowledge that A == **Self** together with the map. Maybe by replacing the map's value type with an enum (Opaque, Concrete, Self).

File: lib/Sema/ConstraintSystem.cpp (hunk 7 above)
Comment (AnthonyLatsis) [reply lib/Sema/ConstraintSystem.cpp]: @slavapestov I've been playing with this part to better understand the impact on the constraint system, and whether I can do something more reasonable than just commenting out existing code.
Having a better picture, I now don't really understand why we need this code at all. 

When the base type is existential, we open and bind it to a type variable as the `Self` generic parameter. In contrast, the archetype acts as `Self` in case of an archetype base.
https://github.com/AnthonyLatsis/swift/blob/master/lib/Sema/ConstraintSystem.cpp#L1351-1353
But then comes this code that replaces `Self` with the existential base when the base is existential. So says the comment, but actually, it replaces the type parameter that represents `Self`, which is bound to the opened existential base, which is an archetype. When we make this replacement, the resulting type becomes free of type variables and thus isn't simplified because `simplifyType` does not accept types without type variables. This is OK hitherto, since not a single existential base with associated types would have tripped over this code, and no associated types means there is nothing to simplify. The replacement in question probably slightly optimizes the CS and constraint generation, but still...
Anyways, existential bases that have concrete associated types are a different story, because these associated types must be simplified to their concrete equivalents. This does not happen without the type variables that continue getting replaced because `simplifyType` isn't called (as it would in case of an archetype *base*). So from Swift's perspective, all that has to be done to make it work is replace, say, `func foo(arg: Proto)` with `func foo<T: Proto>(arg: T)`.
___
I see two approaches to this issue, but the second is mentioned merely for the sake of completeness:
* Remove the code in question, or only resort to these type variable replacements when the base type doesn't have dependent member types (`openedArchetype->getAllNestedTypes().size() == 0` is this efficient though?). 
* Preserve replacements, but replace with the opened archetype instead so that the simplification can take place. Also, we would still need to allow types without types variables in `simplifyType` (not as trivial as removing an `if` statement).

File: test/type/protocol_types.swift (hunk 9 above)
Comment (AnthonyLatsis) [reply test/type/protocol_types.swift]: > Add a SILGen test file to make sure SILGen generates things correctly (take the subset of your 'valid' test cases from test/type/protocol_types.swift and put them in a SILGen test).

Do I simply paste it in? Unfortunately my knowledge on SILGen is yet superficial, so most the tests look cryptic to me.

**Comment by AnthonyLatsis** [comment]
@slavapestov Could you help me out with some grammar? 

*  How come this is allowed?
    ```swift 
    protocol P { typealias E = Int }
    func foo(arg: P) {
      print(arg.E.self)
    }
    ```
* Unlike static and instance members, should concrete members types be allowed to be accessed on both existential metatypes and protocol metatypes? It makes sense, but I want to verify because currently, accessing a concrete `typealias` through an existential metatype [crashes the compiler](https://bugs.swift.org/browse/SR-9781). 
    ```swift
    func foo(meta1: P.Protocol, meta2: P.Type) {
      print(meta1.ConcreteMemberType.self)
      print(meta2.ConcreteMemberType.self) // Crash
    }
    ```

File: lib/AST/GenericSignatureBuilder.cpp (hunk 6 above)
Comment (AnthonyLatsis) [reply lib/AST/GenericSignatureBuilder.cpp]: Turns out there's an obvious problem with this: conflicts like `P1(A == Int) & P2(A == Bool)` will not be catched. Protocol compositions don't have generic environments right? Perhaps there's a more efficient means of creating a generic environment or GSB for a composition by choosing a `Self` archetype and adding a conformance requirement for each protocol?

**Comment by AnthonyLatsis** [comment]
The **TODO** checklist is on the [root post](https://github.com/apple/swift/pull/21576#issue-241524953). There are still some other unrelated tests failing, so I'm going to focus on splitting this into several commits now before continuing with the checklist. Looking forward to your feedback on the comments. It's hard for me to tell how to properly split the change into fully working increments because there are a lot of cascading changes. Hopefully I can split expression and declaration handling, and protocol and composition support too.

**Comment by slavapestov** [comment]
@AnthonyLatsis Your test case crashes for me on both 4.2 and master:

```
Assertion failed: (isa<X>(Val) && "cast<Ty>() argument of incompatible type!"), function cast, file /Users/spestov/src/llvm/include/llvm/Support/Casting.h, line 255.
Stack dump:
0.	Program arguments: /Users/spestov/src/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/bin/swift -frontend -c -primary-file a.swift -target x86_64-apple-darwin18.2.0 -enable-objc-interop -color-diagnostics -module-name a -o /var/folders/wg/nmvf884n4b14lt0yw4khjsv00000gn/T/a-a5c34f.o 
1.	While type-checking 'foo(arg:)' (at a.swift:2:1)
2.	While type-checking statement at [a.swift:2:18 - line:4:1] RangeText="{
  print(arg.E.self)
"
3.	While type-checking expression at [a.swift:3:3 - line:3:19] RangeText="print(arg.E.self"
0  swift                    0x00000001080d9a78 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 40
1  swift                    0x00000001080d8cf5 llvm::sys::RunSignalHandlers() + 85
2  swift                    0x00000001080da082 SignalHandler(int) + 258
3  libsystem_platform.dylib 0x00007fff5f09db3d _sigtramp + 29
4  libsystem_platform.dylib 0x0000000000003e88 _sigtramp + 2700501864
5  libsystem_c.dylib        0x00007fff5ef5b1c9 abort + 127
6  libsystem_c.dylib        0x00007fff5ef23868 basename_r + 0
7  swift                    0x00000001052ffb28 (anonymous namespace)::ExprRewriter::coerceObjectArgumentToType(swift::Expr*, swift::Type, swift::ValueDecl*, swift::AccessSemantics, swift::constraints::ConstraintLocatorBuilder) + 1080
8  swift                    0x00000001052eca8b (anonymous namespace)::ExprRewriter::buildMemberRef(swift::Expr*, swift::Type, swift::SourceLoc, swift::constraints::OverloadChoice, swift::DeclNameLoc, swift::Type, swift::constraints::ConstraintLocatorBuilder, swift::constraints::ConstraintLocatorBuilder, bool, swift::FunctionRefKind, swift::AccessSemantics, bool) + 2475
```

**Comment by AnthonyLatsis** [comment]
> Your test case crashes for me on both 4.2 and master

I see. Must be because I was using a late December build.

> I think so, and I suspect the crashes here can all be fixed together by cleaning up parts of the member access code. Do you mind filing a JIRA? It would be a good starter bug also.

For the type alias on instance? Sure – [SR-9784](https://bugs.swift.org/browse/SR-9784) I'm actually pretty sure I've bumped into the code that was allowing this in my build, somewhere around [here](https://github.com/apple/swift/blob/b422fdcaf711a012944b56ad01d014cc6d0e2953/lib/Sema/CSSimplify.cpp#L3448).

**Comment by slavapestov** [comment]
@AnthonyLatsis perhaps being able to write `Q.A.self` where `A` is an associated type is a bridge too far...

**Comment by davidungar** [comment]
Oh, what a tangled web we weave… :)

**Comment by AnthonyLatsis** [comment]
> @AnthonyLatsis perhaps being able to write Q.A.self where A is an associated type is a bridge too far...

We already can do that, and it works just fine (including `Q.Assoc.self` where `Assoc == Self`). It was a misjudgment of mine to assume this is unprecedented (sorry if I mislead you). I just need to make sure it consistently *works* when an associated type is concrete.

**Comment by AnthonyLatsis** [comment]
> @AnthonyLatsis perhaps being able to write Q.A.self where A is an associated type is a bridge too far...

I think that, just like with type aliases, it's OK as long as `A` is actually a concrete type we can represent.

**Comment by AnthonyLatsis** [comment]
I've split the change into three pieces, doing my best to keep them self-contained. Apart from the nasty special `Assoc == Self` case, which I haven't touched yet, there are two main issues left. I would appreciate your advice on these.

* The first one is about inefficiency of repeatedly creating generic environments for composition types in `resolveNestedType`. The constraint system has a general approach to resolving and dealing with dependent member types - using archetypes. Because we are now allowed to use protocols that specialize their associated types, I *need* archetypes and generic environments to resolve them to their concrete equivalents. Protocols are easy, you use `Self` and `mapTypeIntoContext` to get an archetype. However, protocol compositions don't have generic environments, so I have to create one from scratch from all the protocols in order to resolve an associated type. I could find a protocol that specializes it and use its generic environment, but then *how* do we diagnose `P1(Assoc == Int) & P2(Assoc == Bool)`, and *when*? Maybe protocol compositions, as types, also deserve generic environments? This situation somewhat reminds me of generalized existentials. 

* Because of the potential need to resolve dependent member types, I had to [comment out code that was replacing the type variable bound to the opened *existential* base with the original base](https://github.com/apple/swift/pull/21576#discussion_r244835245). Otherwise, a type that contains `P.Assoc` remains untouched by the CS and this results in either a conversion error or an unhandled coercion assertion. Simply put, this is because `simplifyType` does not let types without type variables pass (you can read more insight on the reasons in response to the comment I linked). The code in question was also desugaring any type aliases along the way through `transform`, but obviously, one cannot simply desugar an associated type. 
The issue is that there are several tests failing due to me commenting out that code. Most of them are crashes on an unhandled coercion, but I haven't looked any closer. I would like to know your opinion on whether this approach makes any sense, or should I better be trying to modify `simplifyType` and derived logic instead so that types like `P.Assoc` also get simplified? 
For clarity, I am pasting a comparative example of a constraint system for a previously unsupported case and its generic equivalent.

<details><summary>Unfold</summary>

```swift
protocol P1 {
  associatedtype Assoc
  func foo(arg: Assoc)
}
protocol P2: P1 where Assoc == Int {}
```
```swift 
func foo(arg: P2) {
  arg.foo(arg: 5)
}
```
```
Score: 0 0 0 0 0 0 0 0 0 0 0
Type Variables:
  $T0 [lvalue allowed] as PP2 @ locator@0x7fc52c050418 [DeclRef@]
  $T1 [lvalue allowed] as (Int) -> () @ locator@0x7fc52c050498 [UnresolvedDot@ -> member]
  $T2 as PP2 @ locator@0x7fc52c0504e0 [UnresolvedDot@ -> member -> generic parameter 'Self']
  $T3 literal=3 bindings={(subtypes of) (default from ExpressibleByIntegerLiteral) Int} @ locator@0x7fc52c050638 [IntegerLiteral@]
  $T4 as () @ locator@0x7fc52c050708 [Call@ -> function result]

Active Constraints:
Inactive Constraints:
  $T3 literal conforms to ExpressibleByIntegerLiteral [[locator@0x7fc52c050638 [IntegerLiteral@]]];
  $T3 arg conv Int [[locator@0x7fc52c0507c0 [Call@ -> apply argument -> comparing call argument #0 to parameter #0]]];
Resolved overloads:
  selected overload set choice PP2.foo: $T1 == ($T2.Assoc) -> ()
  selected overload set choice arg: $T0 == PP2

Opened types:
  locator@0x7fc52c050498 [UnresolvedDot@ -> member] opens τ_0_0 -> $T2

Opened existential types:
  locator@0x7fc52c050498 [UnresolvedDot@ -> member] opens to PP2
```
#### Generic equivalent
```swift
func foo<T: P2>(arg: T) {
  arg.foo(arg: 5)
}
```
```
Score: 0 0 0 0 0 0 0 0 0 0 0
Type Variables:
  $T0 [lvalue allowed] as T @ locator@0x7f845780e018 [DeclRef@]
  $T1 [lvalue allowed] as (Int) -> () @ locator@0x7f845780e098 [UnresolvedDot@ -> member]
  $T2 as T @ locator@0x7f845780e0e0 [UnresolvedDot@ -> member -> generic parameter 'Self']
  $T3 literal=3 bindings={(subtypes of) (default from ExpressibleByIntegerLiteral) Int} @ locator@0x7f845780e238 [IntegerLiteral@]
  $T4 as () @ locator@0x7f845780e308 [Call@ -> function result]

Active Constraints:

Inactive Constraints:
  $T3 literal conforms to ExpressibleByIntegerLiteral [[locator@0x7f845780e238 [IntegerLiteral@]]];
  $T3 arg conv Int [[locator@0x7f845780e3c0 [Call@ -> apply argument -> comparing call argument #0 to parameter #0]]];
Resolved overloads:
  selected overload set choice T.foo: $T1 == ($T2.Assoc) -> ()
  selected overload set choice arg: $T0 == T


Opened types:
  locator@0x7f845780e098 [UnresolvedDot@ -> member] opens τ_0_0 -> $T2
```
</details>

**Comment by AnthonyLatsis** [comment]
@slavapestov I've taken a break, working on some other more trivial issues, but I am still really enthusiastic about finishing this. I'm currently stuck with deciding how the constraint system should behave when handling existentials with concrete associated types.

I know that type aliases don't exist in the constraint system, that is, they always get desugared before being added or so was my observation. The question is whether concrete associated types should also always be simplified before we add them to a CS, or should the resolution be a matter of coercion (`coerceType`). I don't think `getNestedType` belongs to `coerceType`, but I would still like to know your opinion before further digging up existing code.

Right now I do simplify them, but this requires representing them via archetypes even though nothing generic is going on, because `simplifyType` does not accept unprecedented types like `Proto.Assoc`. It can only work with type variables `$T0.Assoc`. In the context of this approach, what do you consider a better solution:
* continue to use archetypes and thus represent concrete associated type bases via type variables (`$T0.Assoc`) or
* rework `simplifyType` and aim to represent concrete ATs using their existential bases (`Proto.Assoc`). 

This is directly related to [this](https://github.com/apple/swift/pull/21576#discussion_r244835245) comment from you, since that is the code that was ensuring existential base usage.

**Comment by AnthonyLatsis** [comment]
OK, so clearly this PR has outlived itself and I feel like I need to start from scratch once I get back to this. @slavapestov Thanks a lot for staying with me! I will let you know as soon as I open a new one.
